File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
# Slide window to check Exact length and Exact count.
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        word_len = len(words[0])
        substr_len = len(words) * word_len
        word_count = Counter(words)
        result = []

        for i in range(word_len):
            left = i
            right = i
            window_count = Counter()

            while right + word_len <= len(s):
                word = s[right:right + word_len]
                right += word_len
                
                if word in word_count:
                    window_count[word] += 1

                    while window_count[word] > word_count[word]:
                        left_word = s[left:left + word_len] # remove the excess word from prev iteration
                        window_count[left_word] -= 1
                        left += word_len

                    if right - left == substr_len:
                        result.append(left)

                else:
                    window_count.clear()
                    left = right
                    
        return result
        

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> result = new ArrayList<>();
        int wordLen = words[0].length(); // 각 단어의 길이
        int concatLen = words.length * wordLen; // 모든 단어를 연결한 문자열의 길이
        Map<String, Integer> wordCount = new HashMap<>();
        
        // 단어의 빈도를 계산
        for (String word : words) {
            wordCount.put(word, wordCount.getOrDefault(word, 0) + 1);
        }

        // 각 offset을 기준으로 슬라이딩 윈도우 검사
        for (int i = 0; i < wordLen; i++) {
            int left = i, right = i;
            Map<String, Integer> windowCount = new HashMap<>();
            
            while (right + wordLen <= s.length()) {
                String word = s.substring(right, right + wordLen);
                right += wordLen;

                if (wordCount.containsKey(word)) {
                    windowCount.put(word, windowCount.getOrDefault(word, 0) + 1);

                    // 단어의 빈도가 초과된 경우, 왼쪽 포인터 이동
                    while (windowCount.get(word) > wordCount.get(word)) {
                        String leftWord = s.substring(left, left + wordLen);
                        windowCount.put(leftWord, windowCount.get(leftWord) - 1);
                        left += wordLen;
                    }

                    // 길이가 맞으면 결과에 추가
                    if (right - left == concatLen) {
                        result.add(left);
                    }
                } else {
                    // 포함되지 않는 단어가 나온 경우, 윈도우 초기화
                    windowCount.clear();
                    left = right;
                }
            }
        }

        return result;
    }
}